build-tools/src/test/groovy/org/elasticsearch/gradle/VersionSpec.groovy

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle
+
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import static org.hamcrest.Matchers.equalTo
+import static org.hamcrest.Matchers.equalTo
+import static org.hamcrest.Matchers.equalTo
+import static org.hamcrest.Matchers.nullValue
+import static org.junit.Assert.assertTrue
+
+@Unroll
+class VersionSpec extends Specification {
+
+    def "parses version string #input with mode #mode"() {
+        expect:
+        assertVersionEquals(input, expectedMajor, expectedMinor, expectedRevision, mode)
+
+        where:
+        input                  | mode                 | expectedMajor | expectedMinor | expectedRevision
+        "7.0.1"                | Version.Mode.STRICT  | 7             | 0             | 1
+        "7.0.1-alpha2"         | Version.Mode.STRICT  | 7             | 0             | 1
+        "5.1.2-rc3"            | Version.Mode.STRICT  | 5             | 1             | 2
+        "6.1.2-SNAPSHOT"       | Version.Mode.STRICT  | 6             | 1             | 2
+        "17.03.11"             | Version.Mode.STRICT  | 17            | 3             | 11
+        "6.1.2"                | Version.Mode.RELAXED | 6             | 1             | 2
+        "6.1.2-SNAPSHOT"       | Version.Mode.RELAXED | 6             | 1             | 2
+        "6.1.2-beta1-SNAPSHOT" | Version.Mode.RELAXED | 6             | 1             | 2
+        "6.1.2-foo"            | Version.Mode.RELAXED | 6             | 1             | 2
+        "6.1.2-foo-bar"        | Version.Mode.RELAXED | 6             | 1             | 2
+        "16.01.22"             | Version.Mode.RELAXED | 16            | 1             | 22
+        "20.10.10+dfsg1"       | Version.Mode.RELAXED | 20            | 10            | 10
+        "2.15"                 | Version.Mode.RELAXED | 2             | 15            | 0
+    }
+
+    def "can use before for comparing"() {
+        expect:
+        assertTrue("1.10.20 is is before 2.0.0", Version.fromString("1.10.20").before("2.0.0"))
```

Comment:
We use `assert` in all the following tests. Which is correct?

Reply:
It would also be nice to have negative tests, i.e. check the behaviour of `#before()` when the versions are the same, or the method argument is not later that the version object. Same for the other methods.